src/plugins/lua/clickhouse.lua

```diff
@@ -741,6 +750,174 @@ local function clickhouse_collect(task)
+  if not params.url then
+    local ip_addr = upstream:get_addr():to_string(true)
+    params.url = connect_prefix .. ip_addr
+  end
+  return rspamd_http.request(params)
+end
+
+local function do_remove_partition(ev_base, cfg, table_name, partition_id)
+  rspamd_logger.debugm(N, ev_base, "removing partition %s.%s", table_name, partition_id)
+  local upstream = settings.upstream:get_upstream_round_robin()
+  local remove_partition_sql = "ALTER TABLE ${table_name} ${remove_method} PARTITION ${partition_id}"
+  local remove_method = (settings.retention.method == 'drop') and 'DROP' or 'DETACH'
+  local sql_params = {
+    ['table_name']     = table_name,
+    ['remove_method']  = remove_method,
+    ['partition_id']   = partition_id
+  }
+
+  local sql = rspamd_lua_utils.template(remove_partition_sql, sql_params)
+
+  local ch_params = {
+    body = sql,
+    ev_base = ev_base,
+    cfg = cfg,
+  }
+
+  if not clickhouse_request(upstream, nil, ch_params) then
+    rspamd_logger.errx(ev_base, "cannot send data to clickhouse server %s: cannot make request",
+            settings['server'])
+  end
+end
+
+local function parse_clickhouse_response(ev_base, cfg, data)
+  rspamd_logger.debugm(N, ev_base, "got clickhouse response: %s", data)
+  if data == nil then
+    -- clickhouse returned no data (i.e. empty resultset): exiting
+    return
+  end
+  local function parse_string(s)
+    local parser = ucl.parser()
+    local res, err = parser:parse_string(s)
+    if not res then
+      rspamd_logger.errx(ev_base, 'Parser error: %s', err)
+      return nil
+    end
+    return parser:get_object()
+  end
+
+  -- iterate over rows
+  local ch_rows = lua_util.str_split(data, "\n")
+  for _, plain_row in pairs(ch_rows) do
+    if plain_row and plain_row:len() > 1 then
+      local parsed_row = parse_string(plain_row)
+      do_remove_partition(ev_base, cfg, parsed_row.table, parsed_row.partition_id)
+    end
+  end
+end
+
+
+--[[
+  nil   - file is not writable, do not perform removal
+  0     - it's time to perform removal
+  <int> - how many seconds wait until next run
+]]
+local function get_last_removal_ago(ev_base)
+  local ts_file = string.format('%s/%s', rspamd_paths['DBDIR'], 'clickhouse_retention_run')
+  local f, err = io.open(ts_file, 'r')
+  local write_file
+  local last_ts
+
+  if err then
+    rspamd_logger.debugm(N, ev_base, 'Failed to open %s: %s', ts_file, err)
+  else
+    last_ts = tonumber(f:read('*number'))
+    f:close()
+  end
+
+  write_file, err = io.open(ts_file, 'w')
+  if err then
+    rspamd_logger.errx(ev_base, 'Failed to open %s, will not perform retention: %s', ts_file, err)
+    return nil
+  end
+
+  local current_ts = os.time()
+
+  if last_ts == nil or (last_ts + settings.retention.period) <= current_ts then
+    local res
+    res, err = write_file:write(tostring(current_ts))
+    if err then
+      rspamd_logger.errx(ev_base, 'Failed to write %s, will not perform retention: %s', ts_file, err)
```

Comment:
ev_base seems to be an invalid userdata here. I'd recommend to use `rspamd_config`: https://github.com/vstakhov/rspamd/blob/master/src/lua/lua_logger.c#L503 (or fix it here)